[클린코드 리액트 3기 이준근] 페이먼츠 미션 Step2

Jxxunnn opened the pull request:
## 프로젝트 실행 방법
프로젝트 시작하기
```bash
npm install && npm run dev
```
스토리북 시작하기
```bash
npm run storybook
```

## 기술 스택

- React
- TypeScript
- Tailwind CSS
- Storybook
- Vite

## 작업 내용

카드 추가 확인
- [x] 이전 폼에서 입력된 카드를 보여준다.
- [x] 카드 별칭을 입력할 수 있다.
  - [x] placeholder는 카드 별칭 (선택)이다.
  - [x] 빈 입력값인 경우, 카드사 이름이 별칭으로 저장된다.
  - [x] 최대 길이는 10자리이다.
  - [x] 확인 버튼을 누르면, 카드 목록 페이지로 이동한다.
 
카드 목록
  - [x] 카드 목록을 조회할 수 있다.
  - [x] 카드 목록은 최신순(내림차순)으로 정렬된다.
  - [x] 목록 최상단에 +을 누르면 카드 추가 페이지로 이동한다.
  - [x] 카드를 클릭하면, 카드 별칭 수정(카드 추가 완료 페이지)로 이동한다.
  - [x] 카드를 삭제할 수 있다.
  
  ## 비고
  
step 2와 step 3이 연결되어 있어서 현재 진행 중인 step 2에서 어느 정도까지 개발해야 하는지 판단하기 어렵네요. 🧐 우선 더 진행하면 코드 라인이 1000자를 넘어가 리뷰가 어려워 질 것으로 예상돼 PR을 제출합니다.

이전 PR에서 페이먼츠 미션에 관련된 궁금증은 모두 해결한 것 같아서 질문은 따로 없습니다. 코드를 기반으로 소통하면 좋을 것 같습니다! 소중한 피드백 기다리고 있겠습니다. 🙏

Jxxunnn committed:
project kick-off: vite + react + ts

Jxxunnn committed:
chore: install and configure tailwind css

Jxxunnn committed:
feat: remove unused style files

Jxxunnn committed:
chore: install libraries

- react-router v6
- tailwind merge

Jxxunnn committed:
feat: config absolute path

Jxxunnn committed:
fix: style global css

Jxxunnn committed:
feat: implement react router

Jxxunnn committed:
feat: implement atom components

- Box
- Button
- Container
- HFlex
- Icon
- MobilePanel
- Portal
- Text
- VFlex

Jxxunnn committed:
feat: implement molecule components

- Header
- IconButton
- Input
- Modal

Jxxunnn committed:
feat: config typography variables

Jxxunnn committed:
feat: implement organism component

- CreditCardInput

Jxxunnn committed:
feat: implement template component

- CreateCardForm

Jxxunnn committed:
chore: add gitignore ongoing works

Jxxunnn committed:
feat: wrap the app with layout component

Jxxunnn committed:
chore: install and configure storybook

Jxxunnn committed:
feat: implement atom component stories

Jxxunnn committed:
feat: implement molecule component stories

Jxxunnn committed:
feat: implement organism component stories

Jxxunnn committed:
feat: implement template component story

Jxxunnn committed:
chore: temporarily removed routing logic

해당 페이지는 현재 스코프에 개발되지 않아, 임시로 alert로 대체하였습니다.

Jxxunnn committed:
feat: write storybook readme

Jxxunnn committed:
chore: update readme

Jxxunnn committed:
chore: rename storybook readme

Jxxunnn committed:
refactor: remove unused routes

Jxxunnn committed:
refactor: rename Box component

Jxxunnn committed:
chore: remove comment

Jxxunnn committed:
chore: remove useless ignore lines

Jxxunnn committed:
refactor: separate and combine typography type

Jxxunnn committed:
feat: pass html button type as rest props

Jxxunnn committed:
chore: config absolute path

Jxxunnn committed:
refactor: move the function to the common utility

Jxxunnn committed:
refactor: replace magic number with constant

Jxxunnn committed:
refactor: remove 'Root' subcomponent

Jxxunnn committed:
refactor: rename 'Input' to 'TextField'

label과 input을 합성하여 사용하는 컴포넌트이기에 primitive ui로 착각할 여지가 있어 이름을 변경합니다.

Jxxunnn committed:
refactor: convert replacer prop to hooks pattern

- UI data와 state의 동기화를 위해 TextField 컴포넌트에서 replacer prop 제거
- replacer prop을 사용하는 컴포넌트를 Hooks Pattern으로 개선
- withReplacer 스토리 제거

Jxxunnn committed:
refactor: remove 'IconButton' component

'Button'컴포넌트로 대체 가능하여 삭제합니다.

Jxxunnn committed:
feat: config absolute path

Jxxunnn committed:
chore: config gitingore

해당 파일들 작업 완료하여 원복합니다.

Jxxunnn committed:
feat: implement molecule components

- TextField variant 추가
- BottomSheet 컴포넌트 추가
- 스토리 작성

Jxxunnn committed:
feat: improve atom components

Jxxunnn committed:
feat: implement organism components

- CreditCard
- CreditCardColorPicker

Jxxunnn committed:
feat: modify label value

Jxxunnn committed:
feat: implement template components

- CardEdit
- CardList
- CardRegistrationSuccess
- CreateCardForm

Jxxunnn committed:
feat: implement credit card entity

Jxxunnn committed:
feat: implement credit card repository

Jxxunnn committed:
feat: implement useOutsideClick hook

Jxxunnn committed:
feat: implement page component 'CardForm'

Jxxunnn committed:
feat: implement page component 'CardList'

Jxxunnn committed:
feat: implement page component 'CardEdit'

Jxxunnn committed:
feat: implement route

Jxxunnn committed:
Merge branch 'jxxunnn' into main

Jxxunnn commented on src/hooks/useOutsideClick.ts:
이 파일은 오픈 소스에서 가져온 코드입니다. 내부 로직은 보지 않으셔도 됩니다! 
ref가 할당된 요소 외부를 클릭했을 때 콜백 함수를 전달하는 역할을 합니다.

igy95 commented on src/App.tsx:
```diff
@@ -1,11 +1,15 @@
 import { Route, BrowserRouter as Router, Routes } from 'react-router-dom';
-import CreateCardForm from './components/templates/CreateCardForm';
+import CardEdit from './components/pages/CardEdit';
+import CardForm from './components/pages/CardForm';
```
'카드'의 정보를 기입하는 '폼' 형식의 컴포넌트는 다른 페이지에서도 충분히 구현될 가능성이 있을 듯하여, 해당 이름은 카드를 생성하는 페이지라는 의미로 수정되어야 할 것 같습니다~

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/135/commits/6d0447fb72686b7012b7c26ae8c261a6657eb793

igy95 commented on src/components/molecules/BottomSheet/index.tsx:
```diff
@@ -0,0 +1,21 @@
+
+interface Props {
+  onClose?: () => void;
+  children: ReactNode;
```
`PropsWithChildren`을 써보면 어떨까요?

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/135/commits/18d1890b01005093a5876e3f0561ac24ba0a74d5

igy95 commented on src/components/molecules/BottomSheet/index.tsx:
```diff
@@ -0,0 +1,21 @@
+import { ReactNode, forwardRef } from 'react';
+
+interface Props {
+  onClose?: () => void;
```
`onClose`는 사용이 안되고 있는 듯 하네요!

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/135/commits/18d1890b01005093a5876e3f0561ac24ba0a74d5

igy95 commented on src/components/organisms/CreditCardColorPicker/index.tsx:
`CreditCardCompanyPicker`가 좀 더 자연스럽지 않을까요~?

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/135/commits/5d48ae68494d82134ca55f8b0929519ac4190148

igy95 commented on src/components/organisms/CreditCardColorPicker/index.tsx:
```diff
@@ -0,0 +1,47 @@
+] as const;
+
+interface Props extends ComponentProps<'div'> {
+  onSubmit: () => void;
```
폼 ui가 아니라서 `onClick`이 어떨까 싶네요

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/135/commits/5d48ae68494d82134ca55f8b0929519ac4190148

igy95 commented on src/components/pages/CardEdit/index.tsx:
```diff
@@ -0,0 +1,51 @@
+  const cardId = params.id;
+  const cardRepository = new CreditCardRepository();
+
+  if (!cardId) {
+    redirect('/');
+    return;
+  }
```
`useEffect` 안쪽이 아닌, 컴포넌트 바로 안에 페이지 이동과 같은 로직이 동기적으로 선언되면 렌더링 과정에서 예상치 못한 사이드 이펙트가 발생될 수 있을 것 같습니다.

Jxxunnn replied:
```
  const params = useParams();
  const cardId = params.id;
  
   if (!cardId) {
    redirect('/');
    return;
  }
```

useParams를 호출하여 얻은 cardId는 컴포넌트와 렌더링 간에 동기화가 되어있는 상태인지라 사이드 이펙트가 발생하지 않을 것으로 예상하고 있습니다. 혹시 어떤 부분에서 사이드 이펙트가 발생할 수 있을까요? 🧐

igy95 replied:
컴포넌트 트리의 root부터 하위 컴포넌트가 존재하지 않을 때까지 호출하는 과정을 '컴포넌트를 그리는 과정'이라고 보았을 때, 컴포넌트를 반복적으로 return 하는 도중에 다른 path로 redirect하는 로직이 끼어있다는 건, 컴포넌트를 미처 다 그리지 못하고 다른 곳으로 이동한다는 의미와 같다고 볼 수 있습니다. 이는 React의 입장에서는 사이드 이펙트에 해당하며, 이러한 이펙트를 다루기 위해 `useEffect`를 사용하는 것을 권장한다고 생각합니다.

또한 조금 먼 생각일 수도 있지만, SSR을 대응하는 경우를 고려한다면 해당 로직은 서버 단에서도 돌아가는 로직으로 볼 수 있기 때문에 예기치 못한 오류를 나타낼 수 있을 거 같아요.

igy95 commented on src/components/pages/CardEdit/index.tsx:
```diff
@@ -0,0 +1,51 @@
+
+  const targetCard = cardRepository.findById(cardId);
+
+  const handleSubmitFinal = (cardNickname: string) => {
```
`Final`이라는 표현이 해당 메서드의 호출 시점에 제약을 주는 것 같네요!

Jxxunnn replied:
함수 내부 로직상 제약을 준 것이 맞습니다!
prop을 전달받은 하위 컴포넌트에서는 이 사실을 모르므로 주어진 함수를 이벤트에 따라 호출하는데 집중할 수 있을 것이라 생각하는데요.  
혹시 어느 부분이 어색하다고 느껴지실까요? 🧐

igy95 replied:
어떤 ui에 대한 클릭 핸들러라는 의미만 담아도 충분하지 않을까 싶어 코멘트 남겼습니다~ ux는 기획에 따라 잦은 수정을 예상할 수 있기 때문에 액션을 수행하는 핸들러에 호출되는 순서까지 함수 이름으로 담고 있어야 하나 싶어서요!

igy95 commented on src/repositories/CreditCardRepository.ts:
`repository`는 어떤 레이어인가요?

Jxxunnn replied:
로컬 스토리지에 접근하는 데이터 액세스 레이어입니다. 페이먼츠 과제에서는 추가적인 레이어가 불필요하다고 판단되어 UI와 직접 연결되어 있습니다.

igy95 replied:
로컬 스토리지를 사용하는 부분만 좀 더 유틸 목적에 맞게 추상화하여 (e.g. `get`, `set`, `remove` / json parsing / error handling) 컴포넌트 단에서 바로 사용하는 게 어떨까 싶습니다. `Repository`라는 네이밍이 로컬 스토리지 접근 + 데이터 연산을 담당하는 레이어라는 의미로 와닿기는 힘들 것 같습니다.

igy95 commented on src/components/templates/CardRegistrationSuccess/index.tsx:
```diff
@@ -0,0 +1,72 @@
+  onEdit?: (cardNickname: string) => void;
+}
+
+export default function CardRegistrationSuccess({ onPrev, onSubmit, onEdit }: Props) {
```
`onSubmit`과 `onEdit`이 한 컴포넌트 안에 있는 것보다 생성과 수정의 flow를 각각 담당하는 컴포넌트에서 최대한 공통의 view를 사용하게끔 처리하는 방향은 어떨까요? 현재는 이 컴포넌트에서 두 가지 이상의 역할을 수행하고 있어 코드의 복잡성이 증가할 가능성이 높아보입니다.

Jxxunnn replied:
넵 회원가입 페이지와 로그인 페이지를 떠올리니 금방 납득이 갔습니다. 제안해주신 대로 컴포넌트를 분리하였습니다. 🙏

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/135/commits/82a6158113259adb68b7f4182d954e8a554b2d63

igy95 commented on src/components/templates/CardList/index.tsx:
Atomic 패턴을 잘 사용하지 않는 이유에 대해 해당 컴포넌트를 예시로 들 수 있을 듯 한데요, `templates`라는 폴더임에도 여기에 속한 컴포넌트에서는 뷰 이외에 더 많은 역할을 하고 있는 것으로 보여 폴더 별 역할 경계가 모호하다는 느낌이 듭니다.

Jxxunnn replied:
네 공감합니다. 현석님이 세션에서 제안해주셔서 도입을 해보았는데요. 아토믹 디자인 패턴 회의론자가 되어가고 있는 중입니다.😅 

혼자 작업하는데도 기준을 명확하게 잡기에는 요구 사항과 구현 방식에 따라 케이스가 너무 많은 것 같고, template과 page의 구분은 현재 리액트의 구현 방식과 너무 동떨어져있는 것 같습니다. 또 organism과 molecule을 구분짓기에는 도메인이라는 용어가 너무 포괄적이고 세분화하기 쉽지 않은 것 같습니다.

실무에서 이게 과연 얼라인이 맞춰질까도 의문이네요.🧐 아토믹 컴포넌트 기반으로 컴포넌트 구조를 만들고자 하는 시도는 현실적으로 어렵고, 변화가 빠른 프론트엔드 씬에서 레거시화되기 쉬울 것 같아요.

아토믹 디자인 패턴을 써보지 않았다면 이러한 단점을 느끼지 못했을 것 같고, 어떤 문제를 발견하기 전에 특정 패턴을 적용한다는 것은 위험하다는 것을 느꼈습니다. 다만 멘탈 모델로서 소통하기에는 정말 좋은 관점인 것 같다고 생각해요. 👍

igy95 replied:
인상 깊은 소감이네요! 러닝 포인트가 있었다면 그 자체로 의미 있는 시도였다고 생각합니다. 👍

igy95 commented on src/components/pages/CardForm/CardFormContext.tsx:
```diff
@@ -0,0 +1,195 @@
+
+type UseCardFormContext = ReturnType<typeof useCardFormContext>;
+
+const initialContextState: UseCardFormContext = {
```
context value의 타입이 nullable함을 교정하기 위한 목적이라면,`useContext`를 호출하는 시점에 값을 가져오려는 컨텍스트가 없다면 에러를 던지는 util hook을 따로 만들어 context value type이 non-null함을 보장해줄 수 있을 듯 합니다.

Jxxunnn replied:
아! 정말 좋은 아이디어네요. 목적에 맞게 개선하였습니다. 👍

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/135/commits/563e0ae7663bc59ed46eacb6eef7484b4c4c5492

igy95 commented on src/components/pages/CardForm/CardFormContext.tsx:
- context 코드 외에 나머지 코드는 모두 분리가 필요해보입니다. 덕스 패턴을 고려했다고 해도 현재 파일 내에서 봐야할 코드가 너무 많습니다.
- `useReducer`가 필요한 이유가 무엇이었을까요? context를 선언하셨다면, 이 컨텍스트 안에서 상태를 다루고 해당 상태나 상태 조작 api가 필요한 컴포넌트에 context value를 내려줬어도 충분하지 않았을까 싶어 여쭤봅니다.

Jxxunnn replied:
> context 코드 외에 나머지 코드는 모두 분리가 필요해보입니다. 덕스 패턴을 고려했다고 해도 현재 파일 내에서 봐야할 코드가 너무 많습니다.

넵 고민하던 부분이었는데 분리하였습니다. 다음 스텝에서 기능이 더 많이 추가될 것으로 예상되어 수정이 원활하도록 덕스 패턴을 고려하였는데요. 리덕스를 안써봐서 이정도 스펙에 보일러플레이트가 이렇게 많아질 줄은 상상도 못했네요. 😅

> useReducer가 필요한 이유가 무엇이었을까요? context를 선언하셨다면, 이 컨텍스트 안에서 상태를 다루고 해당 상태나 상태 조작 api가 필요한 컴포넌트에 context value를 내려줬어도 충분하지 않았을까 싶어 여쭤봅니다.

라이브러리가 제한된 미션 상황 속에서 각 필드의 유효성 검증 실패에 대한 UI 추가가 다음 스텝의 요구사항이었고, 경험 상 state로 각 field의 유효성 상태와 에러메세지 등을 다루면 상태 관리가 어려웠어서 복잡한 상태를 관리하기에 적합한 useReducer를 선택하게 되었습니다.

다만 useReducer 숙련도가 낮아 장점을 잘 취하고 있는지는 잘 모르겠습니다. 광열님은 useReducer를 언제 선택하시나요? 🧐
reducer 코드에 대해서 개선이 필요하다고 생각되는 부분이 있으시다면 부담 없이 알려주세요. 피드백 적극 수용하겠습니다 🙏

Jxxunnn replied:
https://github.com/next-step/react-payments/pull/135/commits/563e0ae7663bc59ed46eacb6eef7484b4c4c5492

igy95 replied:
> 다만 useReducer 숙련도가 낮아 장점을 잘 취하고 있는지는 잘 모르겠습니다. 광열님은 useReducer를 언제 선택하시나요? 🧐
reducer 코드에 대해서 개선이 필요하다고 생각되는 부분이 있으시다면 부담 없이 알려주세요. 피드백 적극 수용하겠습니다 🙏

사용 방식에 대해 딱히 코멘트 드릴 건 없고 상태 관리를 위한 여러 방법론 중에 해당 방식을 사용한 이유가 궁금해 여쭤본 거였습니다 :) 개인적으로 `useReducer`는 잘 사용해본 기억이 없는 것 같습니다. 앞서 말씀 드렸 듯 context provider 내에 로직을 선언해두고 consumer hook을 통해 사용하게 한다던가, 따로 상태관리 툴을 사용해야 한다면 다른 선택지가 더 많다고 보기에 `useReducer`의 위치가 약간 애매하게 느껴지는 게 가장 주된 이유가 아닐까 싶어요.

다만 이 부분은 개인 선호의 영역일 수 있기에 준근님께서 다양하게 도구를 사용해보시면서 스스로의 기준을 확립 시켜보면 좋을 거 같아요!

igy95 reviewed (changes requested):
안녕하세요 준근님! 2단계 미션 코멘트 남겨두었으니 확인 부탁드립니다~

Jxxunnn committed:
refactor: 'BottomSheet' component

- 타입 개선
- 불필요한 prop 제거

Jxxunnn committed:
refactor: 'CreditCardColorPicker' component

- 'CreditCardCompanyPicker'로 이름 변경
해당 컴포넌트의 기능과 이름이 의미적으로 일치하도록 변경
- onClick으로 prop 이름 변경
카드 회사를 선택하는 일반적인 동작을 수행하는 버튼이므로 onClick으로 변경

Jxxunnn committed:
refactor: rename 'CardForm' page

- CardCreate로 변경.
해당 페이지의 목적이 명확하게 표현되어있지 않아 이름을 변경

Jxxunnn committed:
refactor: separate edit and create pages

Jxxunnn committed:
refactor: CardFormContext 개선

- util, reducer, context 코드 분리
- context value의 non-nullable을 보장하는 createProvidedContext 구현

Jxxunnn committed:
Merge branch 'main' of https://github.com/Jxxunnn/react-payments

Jxxunnn commented:
광열님 다시 만나뵙게 되어 영광입니다! 자세한 작업 내용과 의도는 commit body에 남겨놓았습니다. 부족하지만 아무쪼록 잘 부탁드립니다. 많이 배우겠습니다. 감사합니다. 🙏

igy95 reviewed (approved):
안녕하세요 준근님 😀 저도 다시 만나뵙게 되어 반갑네요. 꾸준히 미션 올려주시는 모습 멋집니다 👍 리뷰 반영사항 모두 확인하였고 추가 코멘트 남겨두었으니 확인 부탁드려요! 해당 미션은 이만 머지하도록 하겠습니다.
